File: src/temp/primitives/types/PNone.kt
package temp.primitives.types

import temp.primitives.Primitive

/*
    Used to implement a method that isn't applicable for that class.
*/

class PNone : Primitive() {
    override fun getPrimitiveValue(): Any {
        return PNone()
    }

    override fun negate(): Primitive {
        return PNone()
    }

    override fun not(): Primitive {
        return PNone()
    }

    override fun plus(primitive: Primitive): Primitive {
        return PNone()
    }

    override fun minus(primitive: Primitive): Primitive {
        return PNone()
    }

    override fun times(primitive: Primitive): Primitive {
        return PNone()
    }

    override fun div(primitive: Primitive): Primitive {
        return PNone()
    }

    override fun rem(primitive: Primitive): Primitive {
        return PNone()
    }

    override fun compareGreater(primitive: Primitive): Primitive {
        return PNone()
    }

    override fun compareLesser(primitive: Primitive): Primitive {
        return PNone()
    }
}

File: src/codepulsar/primitive/type/PNone.cpp
#include "PNone.h"


Pulsar::PNone::PNone() {
}

Pulsar::PrimitiveType Pulsar::PNone::getPrimitiveType() {
    return PR_ERROR;
}

bool Pulsar::PNone::isPrimitiveType(PrimitiveType primitiveType) {
    return primitiveType == PR_ERROR;
}

std::any Pulsar::PNone::getPrimitiveValue() {
    return new PNone();
}

Pulsar::Primitive* Pulsar::PNone::unaryNegate() {
    return new PNone();
}

Pulsar::Primitive* Pulsar::PNone::unaryNot() {
    return new PNone();
}

Pulsar::Primitive* Pulsar::PNone::plus(Primitive* primitive) {
    return new PNone();
}

Pulsar::Primitive* Pulsar::PNone::minus(Primitive* primitive) {
    return new PNone();
}

Pulsar::Primitive* Pulsar::PNone::times(Primitive* primitive) {
    return new PNone();
}

Pulsar::Primitive* Pulsar::PNone::div(Primitive* primitive) {
    return new PNone();
}

Pulsar::Primitive* Pulsar::PNone::rem(Primitive* primitive) {
    return new PNone();
}

Pulsar::Primitive* Pulsar::PNone::compareGreater(Primitive* primitive) {
    return new PNone();
}

Pulsar::Primitive* Pulsar::PNone::compareLesser(Primitive* primitive) {
    return new PNone();
}

std::string Pulsar::PNone::toString() {
    return "¯\\_(ツ)_/¯";
}
